src/vs/base/common/event.ts

```diff
@@ -116,6 +116,11 @@ export namespace Event {
 			}
 		});
 
+		const onDidDispose = originalEvent[_onDidDispose];
+		if (typeof onDidDispose === 'function') {
+			once(onDidDispose)(() => emitter.dispose());
+		}
+
```

Comment:
Before, `snapshot` was effectively side-effect free. Now, it subscribes to `onDidDispose` without ever unsubscribing. Isn't this a leak?

I think this should only be done on `onFirstListenerAdd` (and the returned disposable should be cleaned on `onLastListenerRemove`).